### ISSUE TITLE
Successful post processing not removing completed download from SABnzbd

### ISSUE THREAD

**Issue by MrMxyzptlk** [DobyTang/LazyLibrarian#859]
Successful post processing not removing completed download from SABnzbd

File being renamed, moved, and deleted correctly, just not removed from SABnzbd history.
It also shows as processed in LL history.

**Comment by philborman** [comment]
I don't know whether SAB has a "delete from history" option. You can delete a failed/stalled task but not a completed one.

**Comment by MrMxyzptlk** [comment]
Yes, you can always manually delete any 'task' from in SABnzbd.
I thought that LL automatically removed the completed task / job from SABnzbd when post processing?

**Comment by philborman** [comment]
LazyLibrarian only deletes entries from queue if failed. 
Doesn't touch completed history on any of the downloaders.
Just found documentation where SAB does allow removal from history, so could add it. Will look if the other downloaders do too, and maybe make optional.

**Comment by MrMxyzptlk** [comment]
OK, awesome.

**Comment by philborman** [comment]
Seems the one "delete" operation works on deluge (remove from queue and also from history) , and works the same on nzbget, but SAB requires two separate commands. One to delete from queue, and another to delete from history. Not sure about the other downloaders yet. Next version will have the extra sab command in, but I can't test it (don't have a working sab atm) so if you could try with debug on and report back would be useful. Thanks.

**Comment by MrMxyzptlk** [comment]
Would be happy to test. Let me know when it's ready.

**Comment by philborman** [comment]
Just pushed it now, you should see messages in the debug log saying
Request url for (something)   LL(Delete)    and another saying  Request url for (something)   LL(DelHistory)
and each of those should have a Result text from SAB: (hopefully OK)

**Comment by MrMxyzptlk** [comment]
So I started some downloads for books, and they completed.
Rather than wait for the auto post processing I ran a manual PP.
It hung on the following during PP.

2017-06-15 07:32:32	ERROR	Unhandled exception in processDir: Traceback (most recent call last): File "Z:\Applications\lazylibrarian\lazylibrarian\postprocess.py", line 360, in processDir if book_type == 'eBook' and not book_file(pp_path, 'ebook'): File "Z:\Applications\lazylibrarian\lazylibrarian\common.py", line 106, in book_file for fname in os.listdir(search_dir): WindowsError: [Error 3] The system cannot find the path specified: u'Z:\\Applications\\Downloads\\Books\\Finders Keepers (Bill Hodges Trilogy #2) by Stephen King \\*.*'
2017-06-15 07:32:32	DEBUG	Found folder (100%) Z:\Applications\Downloads\Books\Finders Keepers (Bill Hodges Trilogy #2) by Stephen King for eBook Finders Keepers (Bill Hodges Trilogy #2) by Stephen King .epub
2017-06-15 07:32:32	DEBUG	Failed to copy/move file Finders Keepers (Bill Hodges Trilogy #2) by Stephen King .epub to Z:\Applications\Downloads\Books\Finders Keepers (Bill Hodges Trilogy #2) by Stephen King , [Errno 2] No such file or directory: u'Z:\\Applications\\Downloads\\Books\\Finders Keepers (Bill Hodges Trilogy #2) by Stephen King \\Finders Keepers (Bill Hodges Trilogy #2) by Stephen King .epub'
2017-06-15 07:32:32	DEBUG	Skipping Z:\Applications\Downloads\Books\Finders Keepers (Bill Hodges Trilogy #2) by Stephen King, folder is empty
2017-06-15 07:32:32	DEBUG	Skipping Z:\Applications\Downloads\Books\Finders Keepers (Bill Hodges Trilogy #2) by Stephen King, no ebook found
2017-06-15 07:32:32	DEBUG	Found folder (100%) Z:\Applications\Downloads\Books\Finders Keepers (Bill Hodges Trilogy #2) by Stephen King for eBook Finders Keepers (Bill Hodges Trilogy #2) by Stephen King .epub
2017-06-15 07:32:32	INFO	Looking for eBook Finders Keepers (Bill Hodges Trilogy #2) by Stephen King .epub in Z:\Applications\Downloads\Books

**Comment by philborman** [comment]
Interesting. Your debug log looks to be out of order, so a little hard to follow, but it seems windows doesn't like the directory name. Either because it ends in a space character, or maybe the hash. Is # a valid character in a windows directory name? 
Either or both can easily be stripped. At the moment to create the directory name we just take the file name and strip the extension. Unfortunately in this filename there is a space between King and .epub  which might be the cause, or as I said it could be the hash.
I will have a play with directory names and see. Might be if you ask for a directory ending in a space windows refuses, or it strips the space for you?

**Comment by MrMxyzptlk** [comment]
Do not end a file or directory name with a space or a period. Although the underlying file system may support such names, the Windows shell and user interface does not. However, it is acceptable to specify a period as the first character of a name. For example, ".temp".

**Comment by MrMxyzptlk** [comment]
Here is the guide.
https://msdn.microsoft.com/en-us/library/aa365247

**Comment by MrMxyzptlk** [comment]
Removing the space at the end of the filename is letting me continue the process.

**Comment by philborman** [comment]
ok, useful doc. Thanks for that. Seems hash is ok, it's just the space then.  Badly named epub file with a space before the extension, just pushing a fix now...

**Comment by MrMxyzptlk** [comment]
So I see in LL where it is attempting deletion from SABnzbd. But it nevers deletes, and I don't see any indication in the SABnzbd logs that it's being attempted. Here is that log.
[sabnzbd.txt](https://github.com/DobyTang/LazyLibrarian/files/1078027/sabnzbd.txt)

2017-06-15 08:17:48	DEBUG	Deleted James Patterson - Crazy House LL.(34102633), nzb from sabnzbd

**Comment by philborman** [comment]
That's the first delete, delete from queue, there should be a second called
delhistory which runs just after the first one. Do you also have the
message saying "result from sab" as if the first one fails maybe the second
doesn't run?

**Comment by MrMxyzptlk** [comment]
I just did another search and download. I will let you know in the next day or 2 what the results are.
